src/v/cloud_storage/cache_service.cc

```diff
@@ -57,6 +58,68 @@ cache::cache(
   , _cnt(0)
   , _total_cleaned(0) {}
 
+ss::future<>
+cache::recursive_delete_empty_directory(const std::string_view& key) {
+    gate_guard guard{_gate};
+
+    std::filesystem::path canonical_path;
+    try {
+        canonical_path = std::filesystem::canonical(key);
+    } catch (std::filesystem::filesystem_error& e) {
+        if (e.code() == std::errc::no_such_file_or_directory) {
+            vlog(cst_log.debug, "Tried to remove {} that doesn't exist", key);
+            co_return;
+        } else {
+            throw;
+        }
+    }
+
+    std::filesystem::path canonical_cache_dir;
+    try {
+        canonical_cache_dir = std::filesystem::canonical(_cache_dir);
```

Comment:
my fear here is that `std::filesystem::canonical` is making multiple system calls. I'm looking at `vbuild/llvm/src/libcxx/src/filesystem/operations.cpp` and I see `getcwd`, `realpath`, and `pathconf`. if they block we might get reactor stalls.

@ajfabbri @jcsp I think you were looking at very similar topics recently. any thoughts on those and their potential to block? something like `pathconf` is probably reading values out of an in-memory data structures.

@LenaAn do these empty directories need to be immediately removed? an alternative approach is stick the empty directory on a list, and create a timer that periodically wanders around the cache looking for things to garbage collect (and can use that list to be more precise if that's important). 

That would make it easier to move downward in the file system hierarchy and avoid any future problems in which the removal escapes our root directory / file system.

Reply:
the removal is triggered by the timer which runs periodically, and before trying to delete files and directories the timer callback scans the whole cache dir, so it's possible to collect the information about file hierarchy and use it here

Reply:
maybe simply calling [lexically_normal](https://en.cppreference.com/w/cpp/filesystem/path/lexically_normal) will be enough for our purposes (which is just to offer protection against programmer error)